### PR TITLE
Fix some GDScript issues 

### LINKED ISSUE #21366: Typed GDScript - Cannot infer type on iOS
**Godot version:**
From master

**OS/device including version:**
iPhoneX iOS 11.4.1

**Issue description:**
Doing
```
var test: = File.new()
```
Works fine on Windows and MacOS, however when run on an iOS device results in:
```
 **SCRIPT ERROR**: Parse Error: Assigned value does not have a set type, variable type cannot be inferred.
At: res://child.gd:3:GDScript::reload() - Parse Error: Assigned value does not have a set type, variable type cannot be inferred.

```

**Minimal reproduction project:**
[res.zip](https://github.com/godotengine/godot/files/2318364/res.zip)

### LINKED ISSUE #22979: Editor crashing on Linux, and Windows editor built with Linux will crash silently.
## Problem Summary

**Ubuntu 18.04.1 LTS 64-bit and Win10 64-bit v3.1 alpha 43c2868**

**(Also seen in:  fe0db6c4797b8094c64e4ad017e29be866288628)**

The editor crashes when typing a period after a var that contains a class.  It crashes before the auto-completion appears.

The problem goes away when I remove a line from above the line that causes the crash.  I marked it with a comment in the included code below.  In this line I am using  `is` to compare with a preloaded script.

On Unbuntu it produces this stack trace:  [x11.log](https://github.com/godotengine/godot/files/2475516/x11.log)

On Windows, where I first encountered it, it will silently crash with no errors or warnings in the console.

### A period here will crash it.
godot_master_2018-10-13_12-47-22

### Take this line out and no problems.
godot_master_2018-10-13_12-47-12

### Video
You can watch the windows version of this in action at this link -> https://streamable.com/ymr06

## Cross Compile Issue

This one might be elusive, unless it is some problem on X11 builds in general.

It does happen using a 64-bit windows editor, when the editor is built on Linux.  This doesn't happen when I do the same build from Windows using the VS 64-bit compiler.

## How to Reproduce

1. Using Ubuntu 18.04.1 LTS pull the latest the source.
2. Do an editor build with tools=yes bits=64.  (It can also be a Windows build.)
4. Go into the sample project provided below.
5. Enter the script named `error_script.gd`
6. Find the last line with `players.append(anim)`
7. Type a period at the end of `anim.` and if you have the problem it should crash a moment after that.

[3.1 Strange Crash.zip](https://github.com/godotengine/godot/files/2475315/3.1.Strange.Crash.zip)

Also including the scons log of a fresh build, there seems to be a lot of warnings lately, and maybe my builds are setup wrong.

[scons.log](https://github.com/godotengine/godot/files/2475379/scons.log)

### LINKED ISSUE #23520: Type hinted Dictionaries don't accept type hinted nodes as keys.
**Win10 64-bit f84893f**

When type hinting a dictionary and a node, I get an error when I try to use the node as a key for the dictionary.

@vnen Not sure if this is related to something pending.  Or perhaps not a bug.  Though they work, maybe nodes aren't intended to be used as keys?

image

image

[3.1 Dictionary Key Issue.zip](https://github.com/godotengine/godot/files/2546856/3.1.Dictionary.Key.Issue.zip)

### LINKED ISSUE #24464: Parse errors with typed variable & assignment operator (Windows, Export only, Debug OFF)
**Godot version:**
Master dev - c23710adde23a0e415097dee7d8e6f9ba76a0970

**OS/device including version:**
Windows 10

**Issue description:**
The game exported release doesn't run on Windows when a script contains an assignment operator involving a function call and a typed variable.

There's no warning or error in the editor, and it only happens in release builds, which makes the issue difficult to figure out.

**Steps to reproduce:**
1. Create a project with this script:
```func _ready() -> void:
var value1 : float = 1.0;
var value2 : float = 2.0;
value1 *= sign(value2);
```
2. Select Project > Export
3. Add a Windows Desktop preset and leave all options by default
4. Export project
5. **Important**: select Export for Debug **OFF**
6. Run the exported executable, experience quit on startup.
7. Debugging the executable shows parse error on the assignment line:
*Invalid operand types xx & xx to assignment operator '\*'*
(modules/gdscript/gds_parser.cpp, line 7811)

### LINKED ISSUE #24476: File can't be used as static type for function argument
**Godot version:**
3.1.0-alpha3

**OS/device including version:**
windows but probably all

**Issue description:**
File cannot be used in function as optional static type.

**Steps to reproduce:**

```
func _ready():
	var file := File.new()
	test_file(file)

func test_file(file: File)-> void:
	file.close()
```

Once ran, it will fail with an error saying "Invalid type in function 'test_file' in base 'Node2D (my_script.gd)'. Cannot convert argument 1 from Object to Object.". If I remove the type from `test_file(file)` it runs fine. It's possible other types are affected but only experienced on File so far. I tried disabling all the checkboxes under "Project Settings"->"Debug->Gdscript.

### LINKED ISSUE #24493: Static type not detected with new()
**Godot version:**
3.1.alpha-3

**OS/device including version:**
Android (code that works in windows does not work 

**Issue description:**
gdscript types cannot detect static types
it occurs in both member and function variables

**Steps to reproduce:**
```
File f := File.new() # fails with "Parser Error: Assigned value does not have a set type, variable type cannot be inferred."
File f : File = File.new() # works
```
Had similar issues with File, Directory, Tween, AnimationPlayer2D, SpriteFrames when using .new()

### PR DESCRIPTION
**Pull request by vnen**
Fixing some issues with typed GDScript. The comments are on the commit messages.

4d1551dbb: Fix #22979
573fab744: Fix #24464
b0c3a3f2d: Fix #23520
4f72c6be8: Fix #21366, Fix #24493
31433ae8e: Fix #24476

### CODE REVIEW

File: modules/gdscript/gdscript_parser.cpp [hunk 1]
```diff
@@ -6767,10 +6767,10 @@ GDScriptParser::DataType GDScriptParser::_reduce_function_call_type(const Operat
...
-					return_type = original_type;
-					return_type.is_meta_type = false;
-				}
+				return_type = original_type;
+				return_type.is_meta_type = false;
+
+				valid = true; // There's always an initializer, we can asume this is true
 			}
 
 			if (!valid) {
```
Comment (akien-mga) [inline modules/gdscript/gdscript_parser.cpp]: As per the above, this condition could now simply be an `else` branch for the previous `if (is_initializer && original_type.is_meta_type))`. But it's likely better to keep it as is as `valid` is clearer and keeps being used down the road.

File: modules/gdscript/gdscript_parser.cpp [hunk 2]
```diff
@@ -6767,10 +6767,10 @@ GDScriptParser::DataType GDScriptParser::_reduce_function_call_type(const Operat
...
 						default_args_count, is_static, is_vararg);
 
-				if (valid) {
-					return_type = original_type;
-					return_type.is_meta_type = false;
-				}
+				return_type = original_type;
+				return_type.is_meta_type = false;
+
+				valid = true; // There's always an initializer, we can asume this is true
```
[the remark this answers is not in the document]
Comment (akien-mga) [reply modules/gdscript/gdscript_parser.cpp]: It's not a big deal though, so let's merge.

**Comment by akien-mga** [comment]
Thanks!

**Comment by avencherus** [comment]
Nice.  Much appreciated.

**Pull request merged**